src/pipeline/src/etl/processor/vrl.rs

```diff
@@ -0,0 +1,319 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::BTreeMap;
+
+use chrono_tz::Tz;
+use snafu::{OptionExt, ResultExt};
+use vrl::compiler::runtime::Runtime;
+use vrl::compiler::{compile, Program, TargetValue};
+use vrl::diagnostic::Formatter;
+use vrl::prelude::{Bytes, NotNan, TimeZone};
+use vrl::value::{KeyString, Kind, Secrets, Value as VrlValue};
+
+use crate::error::{
+    BytesToUtf8Snafu, CompileVrlSnafu, Error, ExecuteVrlSnafu, FloatNaNSnafu,
+    InvalidTimestampSnafu, KeyMustBeStringSnafu, Result, VrlRegexValueSnafu, VrlReturnValueSnafu,
+};
+use crate::etl::processor::yaml_string;
+use crate::{PipelineMap, Value as PipelineValue};
+
+pub(crate) const PROCESSOR_VRL: &str = "vrl";
+const SOURCE: &str = "source";
+
+#[derive(Debug)]
+pub struct VrlProcessor {
+    source: String,
+    program: Program,
+}
+
+impl VrlProcessor {
+    pub fn new(source: String) -> Result<Self> {
+        let fns = vrl::stdlib::all();
+
+        let compile_result = compile(&source, &fns).map_err(|e| {
+            CompileVrlSnafu {
+                msg: Formatter::new(&source, e).to_string(),
+            }
+            .build()
+        })?;
+
+        let program = compile_result.program;
+
+        // check if the return value is have regex
+        let result_def = program.final_type_info().result;
+        let kind = result_def.kind();
+        if !kind.is_object() {
+            return VrlReturnValueSnafu.fail();
+        }
+        check_regex_output(kind)?;
+
+        Ok(Self { source, program })
+    }
+
+    pub fn resolve(&self, m: PipelineMap) -> Result<PipelineValue> {
+        let pipeline_vrl = m
+            .into_iter()
+            .map(|(k, v)| pipeline_value_to_vrl_value(v).map(|v| (KeyString::from(k), v)))
+            .collect::<Result<BTreeMap<_, _>>>()?;
+
+        let mut target = TargetValue {
+            value: VrlValue::Object(pipeline_vrl),
+            metadata: VrlValue::Object(BTreeMap::new()),
+            secrets: Secrets::default(),
+        };
```

Comment:
In near future, we will need to implement `Target` on our own `PipelineMap`.

Reply:
I also think that the current btreemap doesn't meet our needs anymore, and we should wrap a type to provide more functionality and optimization.